File: app/src/main/java/com/attendance/tracker/ui/components/AttendanceChart.kt
package com.attendance.tracker.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.attendance.tracker.ui.theme.AbsentRed
import com.attendance.tracker.ui.theme.PresentGreen
import com.attendance.tracker.ui.theme.NoClassGray

/**
 * A circular progress indicator that displays attendance percentage
 * visually like a pie chart / donut chart
 */
@Composable
fun AttendancePieChart(
    percentage: Float,
    requiredPercentage: Int,
    modifier: Modifier = Modifier,
    size: Dp = 80.dp,
    strokeWidth: Dp = 8.dp,
    showPercentageText: Boolean = true
) {
    val isAboveRequired = percentage >= requiredPercentage
    val progressColor = if (isAboveRequired) PresentGreen else AbsentRed
    val backgroundColor = NoClassGray.copy(alpha = 0.3f)
    
    Box(
        modifier = modifier.size(size),
        contentAlignment = Alignment.Center
    ) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val strokeWidthPx = strokeWidth.toPx()
            val arcSize = Size(
                width = this.size.width - strokeWidthPx,
                height = this.size.height - strokeWidthPx
            )
            val topLeft = Offset(strokeWidthPx / 2, strokeWidthPx / 2)
            
            // Background circle
            drawArc(
                color = backgroundColor,
                startAngle = 0f,
                sweepAngle = 360f,
                useCenter = false,
                topLeft = topLeft,
                size = arcSize,
                style = Stroke(width = strokeWidthPx, cap = StrokeCap.Round)
            )
            
            // Progress arc
            val sweepAngle = (percentage / 100f) * 360f
            drawArc(
                color = progressColor,
                startAngle = -90f, // Start from top
                sweepAngle = sweepAngle,
                useCenter = false,
                topLeft = topLeft,
                size = arcSize,
                style = Stroke(width = strokeWidthPx, cap = StrokeCap.Round)
            )
            
            // Required percentage indicator line
            val requiredAngle = ((requiredPercentage / 100f) * 360f) - 90f
            val centerX = this.size.width / 2
            val centerY = this.size.height / 2
            val radius = (this.size.width - strokeWidthPx) / 2
            val indicatorLength = strokeWidthPx * 1.5f
            
            val startRadius = radius - indicatorLength / 2
            val endRadius = radius + indicatorLength / 2
            
            val radians = requiredAngle * kotlin.math.PI / 180.0
            val startX = centerX + (startRadius * kotlin.math.cos(radians)).toFloat()
            val startY = centerY + (startRadius * kotlin.math.sin(radians)).toFloat()
            val endX = centerX + (endRadius * kotlin.math.cos(radians)).toFloat()
            val endY = centerY + (endRadius * kotlin.math.sin(radians)).toFloat()
            
            drawLine(
                color = Color.White,
                start = Offset(startX, startY),
                end = Offset(endX, endY),
                strokeWidth = 2.dp.toPx(),
                cap = StrokeCap.Round
            )
        }
        
        if (showPercentageText) {
            Text(
                text = "${percentage.toInt()}%",
                style = MaterialTheme.typography.labelMedium,
                fontWeight = FontWeight.Bold,
                color = progressColor,
                textAlign = TextAlign.Center
            )
        }
    }
}

/**
 * A more detailed attendance visualization with present/absent breakdown
 */
@Composable
fun AttendanceDonutChart(
    presentCount: Int,
    absentCount: Int,
    totalCount: Int,
    modifier: Modifier = Modifier,
    size: Dp = 100.dp,
    strokeWidth: Dp = 12.dp
) {
    val presentPercentage = if (totalCount > 0) (presentCount.toFloat() / totalCount) * 100 else 0f
    val absentPercentage = if (totalCount > 0) (absentCount.toFloat() / totalCount) * 100 else 0f
    
    Box(
        modifier = modifier.size(size),
        contentAlignment = Alignment.Center
    ) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val strokeWidthPx = strokeWidth.toPx()
            val arcSize = Size(
                width = this.size.width - strokeWidthPx,
                height = this.size.height - strokeWidthPx
            )
            val topLeft = Offset(strokeWidthPx / 2, strokeWidthPx / 2)
            
            if (totalCount == 0) {
                // Empty state - gray circle
                drawArc(
                    color = NoClassGray.copy(alpha = 0.3f),
                    startAngle = 0f,
                    sweepAngle = 360f,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(width = strokeWidthPx, cap = StrokeCap.Round)
                )
            } else {
                // Present arc (green)
                val presentSweep = (presentPercentage / 100f) * 360f
                drawArc(
                    color = PresentGreen,
                    startAngle = -90f,
                    sweepAngle = presentSweep,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(width = strokeWidthPx, cap = StrokeCap.Butt)
                )
                
                // Absent arc (red)
                val absentSweep = (absentPercentage / 100f) * 360f
                drawArc(
                    color = AbsentRed,
                    startAngle = -90f + presentSweep,
                    sweepAngle = absentSweep,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(width = strokeWidthPx, cap = StrokeCap.Butt)
                )
            }
        }
        
        Column(
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                text = "${presentPercentage.toInt()}%",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
                color = if (presentPercentage >= 75) PresentGreen else AbsentRed
            )
        }
    }
}

File: iOS/AttendanceTracker/AttendanceTracker/Views/Components/AttendanceChart.swift
import SwiftUI

/// A pie chart showing attendance distribution
struct AttendanceChart: View {
    let presentCount: Int
    let absentCount: Int
    let requiredPercentage: Int
    
    private var totalCount: Int {
        presentCount + absentCount
    }
    
    private var presentPercentage: Double {
        guard totalCount > 0 else { return 0 }
        return Double(presentCount) / Double(totalCount)
    }
    
    private var absentPercentage: Double {
        guard totalCount > 0 else { return 0 }
        return Double(absentCount) / Double(totalCount)
    }
    
    var body: some View {
        HStack(spacing: 16) {
            // Pie Chart
            ZStack {
                if totalCount > 0 {
                    Circle()
                        .stroke(Color.gray.opacity(0.2), lineWidth: 20)
                    
                    // Present arc
                    Circle()
                        .trim(from: 0, to: presentPercentage)
                        .stroke(Color.presentGreen, style: StrokeStyle(lineWidth: 20, lineCap: .round))
                        .rotationEffect(.degrees(-90))
                    
                    // Absent arc
                    Circle()
                        .trim(from: presentPercentage, to: 1)
                        .stroke(Color.absentRed, style: StrokeStyle(lineWidth: 20, lineCap: .round))
                        .rotationEffect(.degrees(-90))
                    
                    // Center text
                    VStack(spacing: 2) {
                        Text(String(format: "%.0f%%", presentPercentage * 100))
                            .font(.headline)
                            .fontWeight(.bold)
                        Text("Present")
                            .font(.caption2)
                            .foregroundColor(.secondary)
                    }
                } else {
                    Circle()
                        .stroke(Color.gray.opacity(0.2), lineWidth: 20)
                    
                    Text("No Data")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }
            .frame(width: 100, height: 100)
            
            // Legend
            VStack(alignment: .leading, spacing: 8) {
                LegendItem(color: .presentGreen, label: "Present", count: presentCount)
                LegendItem(color: .absentRed, label: "Absent", count: absentCount)
                
                Divider()
                
                HStack {
                    Text("Required:")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    Text("\(requiredPercentage)%")
                        .font(.caption)
                        .fontWeight(.medium)
                }
            }
        }
    }
}

struct LegendItem: View {
    let color: Color
    let label: String
    let count: Int
    
    var body: some View {
        HStack(spacing: 8) {
            Circle()
                .fill(color)
                .frame(width: 12, height: 12)
            Text(label)
                .font(.caption)
                .foregroundColor(.secondary)
            Spacer()
            Text("\(count)")
                .font(.caption)
                .fontWeight(.medium)
        }
    }
}

#Preview {
    AttendanceChart(presentCount: 16, absentCount: 4, requiredPercentage: 75)
        .frame(height: 100)
        .padding()
}
